velox/functions/sparksql/tests/MapTest.cpp

```diff
@@ -211,5 +211,17 @@ TEST_F(MapTest, complexTypes) {
        makeArrayVector<int64_t>({{1, 2}, {4, 6}})},
       arrayMapResult1);
 }
+
+TEST_F(MapTest, resultSize) {
+  auto condition = makeFlatVector<int64_t>({1, 2, 3});
+  auto keys = makeFlatVector<int64_t>({3, 2, 1});
+  auto values = makeFlatVector<int64_t>({4, 5, 6});
+  auto mapVector =
+      makeMapVector<int64_t, int64_t>({{{4, 3}}, {{5, 2}}, {{1, 6}}});
+  testMap(
+      "if(greaterthan(c2, 2), map(c0, c1), map(c1, c0))",
```

Comment:
Nice. Would you double check that this test is failing without the fixes from this PR?

You can simplify this test case some more by replacing greaterthan(c2, 2) with c2 and changing condition to be a boolean vector.

Reply:
Sure.
And verified this change fixed the issue.
```
[ RUN      ] MapTest.resultSize
WARNING: Logging before InitGoogleLogging() is written to STDERR
E1129 15:08:16.791761 13853 Exceptions.h:69] Line: ../../velox/vector/ComplexVector.cpp:720, Function:validateArrayVectorBase, Expression: rawOffsets_[i] + rawSizes_[i] - 1 < minChildVectorSize (2 vs. 2) ArrayVectorBase must only point to indices within the base vector's size. Index: 2., Source: RUNTIME, ErrorCode: INVALID_STATE
unknown file: Failure
C++ exception with description "Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: (2 vs. 2) ArrayVectorBase must only point to indices within the base vector's size. Index: 2.
Retriable: False
Expression: rawOffsets_[i] + rawSizes_[i] - 1 < minChildVectorSize
Context: map(c3, c4)
Top-Level Context: switch(greaterthan(c0, 2:BIGINT), map(c1, c2), map(c3, c4))
Function: validateArrayVectorBase
File: ../../velox/vector/ComplexVector.cpp
Line: 720
Stack trace:
# 0  std::shared_ptr<facebook::velox::VeloxException::State const> facebook::velox::VeloxException::State::make<facebook::velox::VeloxException::make(char const*, unsigned long, char const*, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, bool, facebook::velox::VeloxException::Type, std::basic_string_view<char, std::char_traits<char> >)::{lambda(auto:1&)#1}>(facebook::velox::VeloxException::Type, facebook::velox::VeloxException::make(char const*, unsigned long, char const*, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, bool, facebook::velox::VeloxException::Type, std::basic_string_view<char, std::char_traits<char> >)::{lambda(auto:1&)#1})
# 1  facebook::velox::VeloxException::VeloxException(char const*, unsigned long, char const*, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, bool, facebook::velox::VeloxException::Type, std::basic_string_view<char, std::char_traits<char> >)
# 2  facebook::velox::VeloxRuntimeError::VeloxRuntimeError(char const*, unsigned long, char const*, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, bool, std::basic_string_view<char, std::char_traits<char> >)
# 3  void facebook::velox::detail::veloxCheckFail<facebook::velox::VeloxRuntimeError, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>(facebook::velox::detail::VeloxCheckFailArgs const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
# 4  facebook::velox::ArrayVectorBase::validateArrayVectorBase(facebook::velox::VectorValidateOptions const&, int) const
# 5  facebook::velox::MapVector::validate(facebook::velox::VectorValidateOptions const&) const
# 6  facebook::velox::exec::(anonymous namespace)::checkResultInternalState(std::shared_ptr<facebook::velox::BaseVector>&)
# 7  facebook::velox::exec::Expr::eval(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&, facebook::velox::exec::ExprSet const*)
# 8  facebook::velox::exec::SwitchExpr::evalSpecialForm(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)
# 9  facebook::velox::exec::Expr::evalSpecialFormWithStats(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)
# 10 facebook::velox::exec::Expr::evalAllImpl(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)
# 11 facebook::velox::exec::Expr::evalAll(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)
# 12 facebook::velox::exec::Expr::evalWithNulls(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)
# 13 facebook::velox::exec::Expr::evalEncodings(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)
# 14 facebook::velox::exec::Expr::eval(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&, facebook::velox::exec::ExprSet const*)
# 15 facebook::velox::exec::ExprSet::eval(int, int, bool, facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::vector<std::shared_ptr<facebook::velox::BaseVector>, std::allocator<std::shared_ptr<facebook::velox::BaseVector> > >&)
# 16 facebook::velox::exec::ExprSet::eval(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::vector<std::shared_ptr<facebook::velox::BaseVector>, std::allocator<std::shared_ptr<facebook::velox::BaseVector> > >&)
# 17 facebook::velox::functions::test::FunctionBaseTest::evaluate(facebook::velox::exec::ExprSet&, std::shared_ptr<facebook::velox::RowVector> const&, std::optional<facebook::velox::SelectivityVector> const&)
# 18 std::shared_ptr<facebook::velox::BaseVector> facebook::velox::functions::test::FunctionBaseTest::evaluateImpl<facebook::velox::exec::ExprSet>(std::shared_ptr<facebook::velox::core::ITypedExpr const> const&, std::shared_ptr<facebook::velox::RowVector> const&, std::optional<facebook::velox::SelectivityVector> const&)
# 19 facebook::velox::functions::test::FunctionBaseTest::evaluate(std::shared_ptr<facebook::velox::core::ITypedExpr const> const&, std::shared_ptr<facebook::velox::RowVector> const&, std::optional<facebook::velox::SelectivityVector> const&)
# 20 facebook::velox::functions::test::FunctionBaseTest::evaluate(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<facebook::velox::RowVector> const&, std::optional<facebook::velox::SelectivityVector> const&)
# 21 std::shared_ptr<facebook::velox::MapVector> facebook::velox::functions::test::FunctionBaseTest::evaluate<facebook::velox::MapVector>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<facebook::velox::RowVector> const&, std::optional<facebook::velox::SelectivityVector> const&, std::shared_ptr<facebook::velox::Type const> const&)
# 22 void facebook::velox::functions::sparksql::test::(anonymous namespace)::MapTest::testMap<long, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::shared_ptr<facebook::velox::BaseVector>, std::allocator<std::shared_ptr<facebook::velox::BaseVector> > > const&, std::shared_ptr<facebook::velox::BaseVector> const&)
# 23 facebook::velox::functions::sparksql::test::(anonymous namespace)::MapTest_resultSize_Test::TestBody()
# 24 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
# 25 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
# 26 testing::Test::Run()
# 27 testing::TestInfo::Run()
# 28 testing::TestSuite::Run()
# 29 testing::internal::UnitTestImpl::RunAllTests()
# 30 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
# 31 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
# 32 testing::UnitTest::Run()
# 33 RUN_ALL_TESTS()
# 34 main
# 35 __libc_start_main
# 36 _start
" thrown in the test body.
```

Reply:
@zhli1142015 Just to make sure, can you verify that the new test (not the fuzzer) is failing without the fixes in this PR?

Reply:
@zhli1142015 Oh, I see that MapTest.resultSize fails without the fixes. Thanks.